src/internal.rs

```diff
@@ -840,20 +840,108 @@ pub(crate) struct Rav1dFrameData {
 }
 
 impl Rav1dFrameData {
-    pub fn bd_fn(&self) -> &'static Rav1dFrameContextBdFn {
-        let bpc = BPC::from_bitdepth_max(self.bitdepth_max);
-        Rav1dFrameContextBdFn::get(bpc)
+    pub fn assert_has_headers<'a>(&'a self) -> &'a Rav1dFrameDataWithHeaders {
+        if self.seq_hdr.as_ref().is_none() {
+            panic!("required sequence header past this point");
+        }
+
+        if self.frame_hdr.as_ref().is_none() {
+            panic!("required frame header past this point");
+        }
+
+        // Safety: we just asserted that seq_hdr and frame_hdr are Some, and Option<Arc<T>> has the
+        // same layout as Arc<T>. further, `Rav1dFrameData` and `Rav1dFrameDataWithHeaders` have
+        // the same layout with repr(C) ensuring their field ordering is identical.
+        unsafe {
+            std::mem::transmute(self)
+        }
     }
 
-    pub fn frame_hdr(&self) -> &Rav1dFrameHeader {
-        self.frame_hdr.as_ref().unwrap()
+    pub fn assert_has_headers_mut<'a>(&'a mut self) -> &'a mut Rav1dFrameDataWithHeaders {
+        if self.seq_hdr.as_ref().is_none() {
+            panic!("required sequence header past this point");
+        }
+
+        if self.frame_hdr.as_ref().is_none() {
+            panic!("required frame header past this point");
+        }
+
+        // Safety: we just asserted that seq_hdr and frame_hdr are Some, and Option<Arc<T>> has the
+        // same layout as Arc<T>. further, `Rav1dFrameData` and `Rav1dFrameDataWithHeaders` have
+        // the same layout with repr(C) ensuring their field ordering is identical.
+        // Same as `assert_has_headers` overall; we have the sole `&mut self`, so we maintain
+        // exclusivity with the returned `&'a mut` ref returned here.
+        unsafe {
+            std::mem::transmute(self)
+        }
     }
+}
+
+#[repr(C)]
+pub(crate) struct Rav1dFrameDataWithHeaders {
+    pub seq_hdr: Arc<DRav1d<Rav1dSequenceHeader, Dav1dSequenceHeader>>,
+    pub frame_hdr: Arc<DRav1d<Rav1dFrameHeader, Dav1dSequenceHeader>>,
```

Comment:
an obvious first improvement to this patch would be to extract all fields *other than* `seq_hdr` and `frame_hdr` into a struct on its own and share that between both forms of `Rav1dFrameData`